pkg/pillar/types/downloadertypes.go

```diff
@@ -167,3 +168,71 @@ func AllowNonFreePort(gc ConfigItemValueMap, objType string) bool {
 		return false
 	}
 }
+
+// ResolveConfig key/index to this is the ImageID which is allocated by the controller.
+// It will resolve container image tag to sha256
+type ResolveConfig struct {
+	AppID            uuid.UUID
+	ImageID          uuid.UUID
+	DatastoreID      uuid.UUID
+	Name             string
+	NameIsURL        bool // If not we form URL based on datastore info
+	AllowNonFreePort bool
+	PurgeCounter     uint32
+}
+
+// Key : ImageID is used to differentiate different config
+func (config ResolveConfig) Key() string {
+	return fmt.Sprintf("%s-%s-%v", config.AppID.String(), config.ImageID.String(), config.PurgeCounter)
```

Comment:
datastore uuid + name + purgecounter, right?

Reply:
Since the UUID format uses '-', it is much better to use some other character to separate the different parts.
"+", ":" are examples. Note that you can not use '/' since that isn't allowed in a filename.

Reply:
Resolved.